src/preprocess/mod.rs

```diff
@@ -1 +1,17 @@
-pub mod links;
+pub use self::links::LinkPreprocessor;
+
+mod links;
+
+use book::Book;
+use errors::*;
+
+use std::path::PathBuf;
+
+pub struct PreprocessorContext {
+    pub src_dir: PathBuf
+}
+
+pub trait Preprocessor {
+    fn name(&self) -> &str;
+    fn run(&self, ctx: &PreprocessorContext, book: &mut Book) -> Result<()>;
```

Comment:
I'm tossing up whether to include the `Book` as part of the `PreprocessorContext`. Do you think the book to be processed should be part of the context or should we keep it as a second argument?

Reply:
I think it's helpful to distinguish the fact that you only expect the `Book` itself to be mutated, and that the `PreprocessorContext` is additional information it can use to help in that mutation. Having a single parameter `&mut PreprocessorContext` obscures that slightly and has fewer guarantees about what the `Preprocessor` is allowed to do to the context.

Reply:
Ah yeah that's a good point. I completely forgot the `Book` is `&mut` while we pass an immutable reference for the `PreprocessorContext`.